sharktank/sharktank/ops/sharded_impls.py

```diff
@@ -1051,11 +1051,46 @@ def mean_replicated(
     keepdim: bool,
     *,
     dtype: torch.dtype,
-) -> None:
+) -> ReplicatedTensor:
     shards = [mean(shard, dim=dim, keepdim=keepdim, dtype=dtype) for shard in x.shards]
     return ReplicatedTensor(ts=shards)
 
 
+@mean.override(SplitPrimitiveTensor)
+def mean_split(
+    x: SplitPrimitiveTensor,
+    dim: Union[int, List[int]],
+    keepdim: bool,
+    *,
+    dtype: torch.dtype,
+) -> SplitPrimitiveTensor | ReplicatedTensor:
+    if dim != x.shard_dim:
+        # If keepdim == False dim is smaller than shard_dim we need to offset
+        # shard_dim_new by -1 to have it point to the same dimension.
+        shard_dim_new = x.shard_dim
+        if (not keepdim) and dim < x.shard_dim:
+            shard_dim_new -= 1
+
+        shards = [
+            mean(shard, dim=dim, keepdim=keepdim, dtype=dtype) for shard in x.shards
+        ]
+        return SplitPrimitiveTensor(ts=shards, shard_dim=shard_dim_new)
+    else:
+        gathered = cat(
+            [
+                (
+                    transfer_to_logical_device(shard, x.devices[0])
+                    if i != 0
+                    else barrier_on_logical_device(shard, x.devices[0])
+                )
+                for i, shard in enumerate(x.shards)
+            ],
+            dim=dim,
+        )
+        meaned = mean(gathered, dim=dim, keepdim=keepdim, dtype=dtype)
+        return ReplicatedTensor(ts=meaned, shard_count=x.shard_count, devices=x.devices)
```

Comment:
It is not critical for correctness, but for performance reasons I think we could change the implementation to
1. Each device reduces along dim.
2. All-reduce across the devices.
3. Divide by the total number of elements in the above reductions.

For most cases this should bring down communication a lot.

Reply:
This can be in a later PR.

Reply:
I considered this, my worry was about potential under/over flow. And that underlying floating point operations would no longer be the same leading to slight differences between the results.

Reply:
The operation does not guarantee in what order are additions performed. The underlying implementation may tile on a reduction dimension even on a GPU. Actually, tiling would yield better numerical stability. The accumulator would not get that big and lose precision when adding small and big numbers together.